[Deepin-Kernel-SIG] [linux 6.6-y] [WangXun] deepin: net: txgbe: add support  for wangxun 10G/25G/40G

melooop opened the pull request:
1.net: wangxun: txgbe: add support for wangxun 10G/25G/40G
2.net: wangxun: remove wangxun txgbe inbox drivers
3.net: txgbe: fix invoke hot reset multi times unexpectedly
4.net: txgbe: fix race condition when recover or close/suspend
5.net: txgbe: fix fec mode issues
6.net: txgbe: set pf promisc and vlan cannot receive packets
7.net: txgbe: fix 25G link issues

all patches make deepin supporting wangxun 10G/25G/40G NICs.

Commit:
deepin: net: wangxun: txgbe: add support for wangxun 10G/25G/40G

wangxun inclusion
category: feature

this driver is supported for wangxun SP1000A/WX5025/WX5040 chips,
and supports TSO, tx checksum, rx checksum, RSS, flow director, vlan
insert, vlan strip, vlan filter offload.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix DAC cable link down

wangxun inclusion
category: feature

where plugging and unplugging the DAC cable during compilation
with force mode might not link, and tx ffe parameter and bypassctle
need to be modified accordingly.
After polling fec, modify the fec mode to RS if supported, and adding
an interrupt to the phy enables the local end to receive interrupt
synchronization configuration when plugging or unplugging the dac
from the opposite end.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix pktgen cannot stop

wangxun inclusion
category: feature

When using pktgen to send packet with count argument,  the pktgen cmd
cannot quit successfully and the CPU utilization would increase to 100%.
The bug is introduced by delayed-releasing tx skbs/buffers mechanism.
In some working scenarios, trigger irq per 2 seconds is a bit low,
increase the trigger frequency to per 250ms.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix missing PFC mask configuration on AML

wangxun inclusion
category: feature

when set 'lldptool -T -i dev PFC enable=1,2,4', pfc priority0 still
cause flow control. So add write pfc_mask(0x1f6c[0:7]) on amlite
to isolate disabled priorities.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix VF multicast address removal not taking effect

wangxun inclusion
category: feature

VF continued reveiving multicast traffic for removed addresses,
regardless of being bound to DPDK or Linux kernel driver.

PF driver mailbox handling for VF update-multicast cmd only re-added
addresses from the message buffer, failing to remove stale entries.
Unlike PF's own multicast updates (which flush/reload the entire table),
VF commands left old addresses intact.

To solve this bug, we modify PF multicast update flow: flush entire
table when processing VF mailbox commands and then repopulate table
by iterating through all VF's multicast addresses.

Known limitations about VF multicast:

1) Shared mcast table between PF and all VFs: When any VF configures
a multicast address, if other VFs have multicast promiscuous enabled,
they will also receive multicast packets destined for that address.

2) Ineffective address deletion: When attempting to delete a multicast
address, if either the PF or other VFs have configured this same address,
the deletion operation will not actually take effect.

3) Hash collision possibility: The multicast address table uses only
12 bits of the multicast MAC address for bitmap mapping, while multicast
addresses contain 23 significant bits. This mismatch may cause different
addresses mapped to the same table entry.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix bond port rate is unknown when linkup

wangxun inclusion
category: feature

For 25G, when using the dual-speed optical module to link to
10G, the driver first reported the link up and then re-attempted
to configure the phy to 25G, resulting in the bond port being
unable to obtain the speed at this time.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix can't read rx_pb_n_pxoff

wangxun inclusion
category: feature

after turn on dcb and set pfc, rx_pb_n_pxoff(n=1~7) will keep be 0
after receive pfc packet. becaus amlite use 0x01FA0+ [0..7 * 0x4]
and 0x01FC0+ [0..7 * 0x4] record rx_pb_n_pxoff and rx_pb_n_pxon.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix i2c issues

wangxun inclusion
category: feature

In txgbe, when ethtool -m port will return -EINVAL with DAC
because not record i2c info when identify sfp. And If repeatly
reinit_locked, when txgbe_down(), may read i2c error in record
i2c info flow. So skip it when port down.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: change default tx queue depth

wangxun inclusion
category: feature

ethtool -g ethx tx 128, and then tx ring tx timeout.
because amblite reserved 192 desc entry, but ring depth
set to 128 is less than reserved num.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix 25G link issues

wangxun inclusion
category: feature

Currently, we hope to link to rs fec, but when switching to fec
at the end, we didn't wait, which led to the can not link back
to 25G during the dual-speed switch. So, when switching to RS fec,
wait for a while for the link.
And after ethtool -s ethx speed maxspeed duplex full autoneg off
and down up port, then ethtool -s enp1s0f0 speed maxspeed
duplex full autoneg off, can't set port speed in multispeed sfp.
Because when port up, the phy.autoneg_advertised reinit to default
full speed but autoneg not reinit.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: set pf promisc and vlan cannot receive packets

wangxun inclusion
category: feature

e.g.
1.echo 1>/sys/class/net/enp25s0f0/device/sriov numvfs;
2.ip link set dev enp25s0f0 vf 0 trust on;
3.ifconfig enp25s0f0 promisc;
4.ip link add link enp25s0f0 name enp25s0f0.101 type vlan id 101;

then can't receive vlan 100 packet. because vfta has been wrongly
setting when add vlan. So fix it and delete unused vft_shadow.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix fec mode issues

wangxun inclusion
category: feature

fix can't link 10g when fec mismatch peer in multispeed,
amblit chip in multispeed, if set local fec mismatch peer.
may link no, because phy setting flow error and cannot link
when change fec to auto from rs with baser peer.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix race condition when recover or close/suspend

wangxun inclusion
category: feature

when do pcie recover and close or xdp_setup at same time,
txgbe_close() may be called in txgbe_io_error_detected again.
which ends up in a double free WARN and/or a BUG in free_msi_irqs().
To handle this situation we extend the rtnl_lock() to protect the
call to netif_device_detach() and check for netif_device_present()
to avoid clearing the interrupts second time in txgbe_close();
when do recover, tx_ring->desc be free,
txgbe_tx_queue_clear_error_task in service_task touch tx_ring->desc
will occur the calltrace, so add judgement in
txgbe_tx_queue_clear_error_task.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: txgbe: fix invoke hot reset multi times unexpectedly

wangxun inclusion
category: feature

e.g. when lan0 invoke hot reset function, when
down lan0 but pending bit, it will cause hot reset
again. add io_err flag to avoid it. And if lan1
detected error and raise hot reset flag,
and then lan0 invoke hot reset function and dump
lan0 debug info, but we need lan1 debug info.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

Commit:
deepin: net: wangxun: remove wangxun txgbe inbox drivers

wangxun inclusion
category: feature

OLK-6.6 had wangxun txgbe drivers which is from
linux upstream. But this driver do not have many
capabilities and is limited by the integration strategy
of upstream. Wangxun want to merge new drivers, so remove
it.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

opsiff commented on drivers/net/ethernet/wangxun/Kconfig:
```diff
@@ -38,20 +38,9 @@ config NGBE
-	tristate "Wangxun(R) 10GbE PCI Express adapters support"
+	tristate "Wangxun(R) 10/25/40GbE PCI Express adapters support"
 	depends on PCI
-	depends on COMMON_CLK
```
这些depend和select需要保留吧 否则是否存在选中TXGBE但是不能使用的情况?

melooop replied:
不需要了，这是我们合入上游社区的版本需要的，因为上游社区的诸多限制，导致很多功能需要依赖内核其他模块而不让wangxun驱动自行实现，从而造成需要多个select的情况，在第一个补丁里，我是直接删除了上游社区的驱动。

opsiff replied:
> 不需要了，这是我们合入上游社区的版本需要的，因为上游社区的诸多限制，导致很多功能需要依赖内核其他模块而不让wangxun驱动自行实现，从而造成需要多个select的情况，在第一个补丁里，我是直接删除了上游社区的驱动。

有些编译错误 需要用clang编译一下看看怎么解决 2025-12-03T07:20:40.8258736Z drivers/net/ethernet/wangxun/txgbe/txgbe_hw.c:6449:2: error: unannotated fall-through between switch labels [-Werror,-Wimplicit-fallthrough]
2025-12-03T07:20:40.8263785Z  6449 |         case 0x7F:
2025-12-03T07:20:40.8266605Z       |         ^
2025-12-03T07:20:40.8315789Z drivers/net/ethernet/wangxun/txgbe/txgbe_hw.c:6449:2: note: insert 'break;' to avoid fall-through
2025-12-03T07:20:40.8317087Z  6449 |         case 0x7F:
2025-12-03T07:20:40.8317926Z       |         ^
2025-12-03T07:20:40.8318418Z       |         break; 
2025-12-03T07:20:40.8319908Z drivers/net/ethernet/wangxun/txgbe/txgbe_hw.c:6461:2: error: unannotated fall-through between switch labels [-Werror,-Wimplicit-fallthrough]
2025-12-03T07:20:40.8321376Z  6461 |         case TXGBE_ATR_L4TYPE_MASK:
2025-12-03T07:20:40.8322063Z       |         ^
2025-12-03T07:20:40.8323035Z drivers/net/ethernet/wangxun/txgbe/txgbe_hw.c:6461:2: note: insert 'break;' to avoid fall-through
2025-12-03T07:20:40.8324176Z  6461 |         case TXGBE_ATR_L4TYPE_MASK:
2025-12-03T07:20:40.8324896Z       |         ^
2025-12-03T07:20:40.8325405Z       |         break; 
2025-12-03T07:20:40.8491761Z 2 errors generated.
2025-12-03T07:20:40.8492468Z   CC [M]  net/netfilter/nft_socket.o
2025-12-03T07:20:40.8697506Z make[7]: *** [scripts/Makefile.build:243：drivers/net/ethernet/wangxun/txgbe/txgbe_hw.o] 错误 1
2025-12-03T07:20:40.8700651Z make[7]: *** 正在等待未完成的任务....
2025-12-03T07:20:40.8726644Z   CC [M]  drivers/infiniband/hw/irdma/i40iw_hw.o
2025-12-03T07:20:40.8729896Z   CC [M]  drivers/infiniband/ulp/iser/iscsi_iser.o
2025-12-03T07:20:40.8747653Z   CC [M]  drivers/staging/rtl8723bs/hal/sdio_halinit.o
2025-12-03T07:20:40.8827899Z   CC [M]  drivers/gpu/drm/mwv207/mwv207_sched.o
2025-12-03T07:20:40.8859969Z drivers/net/ethernet/wangxun/txgbe/txgbe_ethtool.c:799:8: error: use of bitwise '&' with boolean operands [-Werror,-Wbitwise-instead-of-logical]
2025-12-03T07:20:40.8861787Z   799 |                         if (ethtool_link_ksettings_test_link_mode(cmd, advertising,
2025-12-03T07:20:40.8862722Z       |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8863580Z   800 |                                                                   10000baseKR_Full) &
2025-12-03T07:20:40.8864325Z       |                                                                   ~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8873709Z   801 |                             ethtool_link_ksettings_test_link_mode(cmd, advertising,
2025-12-03T07:20:40.8874699Z       |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8875454Z   802 |                                                                   1000baseKX_Full) &
2025-12-03T07:20:40.8876227Z       |                                                                   ~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8876944Z       |                                                                                    &&
2025-12-03T07:20:40.8877800Z   803 |                             ethtool_link_ksettings_test_link_mode(cmd, advertising,
2025-12-03T07:20:40.8878607Z       |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8879522Z   804 |                                                                   10000baseKX4_Full))
2025-12-03T07:20:40.8880286Z       |                                                                   ~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8881414Z ./include/linux/ethtool.h:221:2: note: expanded from macro 'ethtool_link_ksettings_test_link_mode'
2025-12-03T07:20:40.8882699Z   221 |         test_bit(ETHTOOL_LINK_MODE_ ## mode ## _BIT, (ptr)->link_modes.name)
2025-12-03T07:20:40.8883456Z       |         ^
2025-12-03T07:20:40.8884171Z ./include/linux/bitops.h:60:29: note: expanded from macro 'test_bit'
2025-12-03T07:20:40.8885182Z    60 | #define test_bit(nr, addr)              bitop(_test_bit, nr, addr)
2025-12-03T07:20:40.8885892Z       |                                         ^
2025-12-03T07:20:40.8886743Z ./include/linux/bitops.h:48:2: note: expanded from macro 'bitop'
2025-12-03T07:20:40.8888367Z    48 |         ((__builtin_constant_p(nr) &&                                   \
2025-12-03T07:20:40.8889759Z       |         ^
2025-12-03T07:20:40.8890976Z drivers/net/ethernet/wangxun/txgbe/txgbe_ethtool.c:799:8: note: cast one or both operands to int to silence this warning
2025-12-03T07:20:40.8892507Z ./include/linux/ethtool.h:221:2: note: expanded from macro 'ethtool_link_ksettings_test_link_mode'
2025-12-03T07:20:40.8893765Z   221 |         test_bit(ETHTOOL_LINK_MODE_ ## mode ## _BIT, (ptr)->link_modes.name)
2025-12-03T07:20:40.8894525Z       |         ^
2025-12-03T07:20:40.8895265Z ./include/linux/bitops.h:60:29: note: expanded from macro 'test_bit'
2025-12-03T07:20:40.8896233Z    60 | #define test_bit(nr, addr)              bitop(_test_bit, nr, addr)
2025-12-03T07:20:40.8897001Z       |                                         ^
2025-12-03T07:20:40.8897814Z ./include/linux/bitops.h:48:2: note: expanded from macro 'bitop'
2025-12-03T07:20:40.8898772Z    48 |         ((__builtin_constant_p(nr) &&                                   \
2025-12-03T07:20:40.8899444Z       |         ^
2025-12-03T07:20:40.8900836Z drivers/net/ethernet/wangxun/txgbe/txgbe_ethtool.c:799:8: error: use of bitwise '&' with boolean operands [-Werror,-Wbitwise-instead-of-logical]
2025-12-03T07:20:40.8902420Z   799 |                         if (ethtool_link_ksettings_test_link_mode(cmd, advertising,
2025-12-03T07:20:40.8903275Z       |                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8904078Z   800 |                                                                   10000baseKR_Full) &
2025-12-03T07:20:40.8904813Z       |                                                                   ~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8905551Z       |                                                                                     &&
2025-12-03T07:20:40.8906438Z   801 |                             ethtool_link_ksettings_test_link_mode(cmd, advertising,
2025-12-03T07:20:40.8907309Z       |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8908135Z   802 |                                                                   1000baseKX_Full) &
2025-12-03T07:20:40.8908858Z       |                                                                   ~~~~~~~~~~~~~~~~
2025-12-03T07:20:40.8909884Z ./include/linux/ethtool.h:221:2: note: expanded from macro 'ethtool_link_ksettings_test_link_mode'
2025-12-03T07:20:40.8911058Z   221 |         test_bit(ETHTOOL_LINK_MODE_ ## mode ## _BIT, (ptr)->link_modes.name)
2025-12-03T07:20:40.8918313Z       |         ^
2025-12-03T07:20:40.8919101Z ./include/linux/bitops.h:60:29: note: expanded from macro 'test_bit'
2025-12-03T07:20:40.8920038Z    60 | #define test_bit(nr, addr)              bitop(_test_bit, nr, addr)
2025-12-03T07:20:40.8920747Z       |                                         ^
2025-12-03T07:20:40.8921490Z ./include/linux/bitops.h:48:2: note: expanded from macro 'bitop'
2025-12-03T07:20:40.8922419Z    48 |         ((__builtin_constant_p(nr) &&                                   \
2025-12-03T07:20:40.8923137Z       |         ^
2025-12-03T07:20:40.8924187Z drivers/net/ethernet/wangxun/txgbe/txgbe_ethtool.c:799:8: note: cast one or both operands to int to silence this warning
2025-12-03T07:20:40.8925690Z ./include/linux/ethtool.h:221:2: note: expanded from macro 'ethtool_link_ksettings_test_link_mode'
2025-12-03T07:20:40.8926910Z   221 |         test_bit(ETHTOOL_LINK_MODE_ ## mode ## _BIT, (ptr)->link_modes.name)
2025-12-03T07:20:40.8927621Z       |         ^
2025-12-03T07:20:40.8928323Z ./include/linux/bitops.h:60:29: note: expanded from macro 'test_bit'
2025-12-03T07:20:40.8929305Z    60 | #define test_bit(nr, addr)              bitop(_test_bit, nr, addr)
2025-12-03T07:20:40.8930045Z       |                                         ^
2025-12-03T07:20:40.8930853Z ./include/linux/bitops.h:48:2: note: expanded from macro 'bitop'
2025-12-03T07:20:40.8932250Z    48 |         ((__builtin_constant_p(nr) &&                                   \
2025-12-03T07:20:40.8932945Z       |         ^
2025-12-03T07:20:40.8957074Z   LD [M]  drivers/scsi/bfa/bfa.o
2025-12-03T07:20:40.8958114Z   CC [M]  drivers/gpu/drm/gma500/intel_gmbus.o
2025-12-03T07:20:40.8989119Z   CC [M]  drivers/infiniband/sw/rdmavt/cq.o
2025-12-03T07:20:40.9394237Z   CC [M]  drivers/infiniband/ulp/ipoib/ipoib_multicast.o
2025-12-03T07:20:40.9826131Z 2 errors generated.
2025-12-03T07:20:40.9867379Z   CC [M]  drivers/platform/x86/intel_scu_pltdrv.o
2025-12-03T07:20:40.9995473Z make[7]: *** [scripts/Makefile.build:243：drivers/net/ethernet/wangxun/txgbe/txgbe_ethtool.o] 错误 1
2025-12-03T07:20:40.9996718Z   CC [M]  drivers/infiniband/hw/irdma/i40iw_if.o

Commit:
deepin: net: txgbe: fix clang compile error

wangxun inclusion
category: bugfix

fix clang [-Werror,-Wimplicit-fallthrough] and
[-Werror,-Wbitwise-instead-of-logical] compile error.

Signed-off-by: Duanqiang Wen <duanqiangwen@net-swift.com>

opsiff commented:
@melooop 更新之后有新的clang编译报错 建议本地先用CC=clang编译测试一下 2025-12-03T13:09:39.6204618Z drivers/net/ethernet/wangxun/txgbe/txgbe_e56.c:3456:4: error: variable 'adapter' is uninitialized when used here [-Werror,-Wuninitialized]
2025-12-03T13:09:39.6234459Z  3456 |                         e_info(drv, "ERROR: Wait E56PHY_CTRL_FSM_RX_STAT_0_ADDR Timeout!!!\n");
2025-12-03T13:09:39.6236172Z       |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2025-12-03T13:09:39.6238948Z drivers/net/ethernet/wangxun/txgbe/txgbe.h:1293:30: note: expanded from macro 'e_info'
2025-12-03T13:09:39.6240441Z  1293 |         netif_info(adapter, msglvl, adapter->netdev, format, ## arg)
2025-12-03T13:09:39.6241326Z       |                                     ^~~~~~~
2025-12-03T13:09:39.6242281Z ./include/net/net_debug.h:109:32: note: expanded from macro 'netif_info'
2025-12-03T13:09:39.6243315Z   109 |         netif_level(info, priv, type, dev, fmt, ##args)
2025-12-03T13:09:39.6244078Z       |                                       ^~~
2025-12-03T13:09:39.6245068Z ./include/net/net_debug.h:93:18: note: expanded from macro 'netif_level'
2025-12-03T13:09:39.6246069Z    93 |                 netdev_##level(dev, fmt, ##args);               \
2025-12-03T13:09:39.6246851Z       |                                ^~~
2025-12-03T13:09:39.6248270Z drivers/net/ethernet/wangxun/txgbe/txgbe_e56.c:3299:31: note: initialize the variable 'adapter' to silence this warning
2025-12-03T13:09:39.6250366Z  3299 |         struct txgbe_adapter *adapter;
2025-12-03T13:09:39.6251078Z       |                                      ^
2025-12-03T13:09:39.6251744Z       |                                       = NULL
2025-12-03T13:09:39.6264454Z   CC [M]  sound/soc/codecs/max98520.o
2025-12-03T13:09:39.6315702Z   CC [M]  drivers/mtd/maps/intel_vr_nor.o
2025-12-03T13:09:39.6338434Z 1 error generated.
2025-12-03T13:09:39.6545077Z make[7]: *** [scripts/Makefile.build:243：drivers/net/ethernet/wangxun/txgbe/txgbe_e56.o] 错误 1

Avenger-285714 commented:
/ok-to-test

opsiff committed:
Merge branch 'linux-6.6.y' into linux-6.6.y
